File: scala/com/mounish/DataStructures/SinglyLinkedList.scala
package com.mounish.DataStructures

/**
 * Singly Linked List implementation in Scala with improvements:
 * - Immutable and mutable versions
 * - Functional programming approach
 * - Pattern matching for cleaner code
 * - Tail recursion for better performance
 * - Type parameterization for generics
 */

// Immutable version (functional approach)
sealed trait LinkedList[+A] {
  def isEmpty: Boolean
  def head: A
  def tail: LinkedList[A]

  def prepend[B >: A](elem: B): LinkedList[B] = Node(elem, this)

  def append[B >: A](elem: B): LinkedList[B] = this match {
    case Empty => Node(elem, Empty)
    case Node(h, t) => Node(h, t.append(elem))
  }

  def delete[B >: A](elem: B): LinkedList[B] = this match {
    case Empty => Empty
    case Node(h, t) if h == elem => t
    case Node(h, t) => Node(h, t.delete(elem))
  }

  def count: Int = this match {
    case Empty => 0
    case Node(_, t) => 1 + t.count
  }

  def printAll(): Unit = {
    def loop(list: LinkedList[A]): Unit = list match {
      case Empty => println()
      case Node(h, Empty) => println(h)
      case Node(h, t) =>
        print(s"$h -> ")
        loop(t)
    }
    loop(this)
  }

  def get(index: Int): Option[A] = {
    @scala.annotation.tailrec
    def loop(list: LinkedList[A], i: Int): Option[A] = list match {
      case Empty => None
      case Node(h, _) if i == 0 => Some(h)
      case Node(_, t) => loop(t, i - 1)
    }
    loop(this, index)
  }

  def toList: List[A] = this match {
    case Empty => Nil
    case Node(h, t) => h :: t.toList
  }
}

case object Empty extends LinkedList[Nothing] {
  def isEmpty: Boolean = true
  def head: Nothing = throw new NoSuchElementException("head of empty list")
  def tail: LinkedList[Nothing] = throw new NoSuchElementException("tail of empty list")
}

case class Node[A](head: A, tail: LinkedList[A]) extends LinkedList[A] {
  def isEmpty: Boolean = false
}

// Mutable version (similar to Java implementation)
class MutableSinglyLinkedList[A] {
  private class ListNode(var data: A, var next: Option[ListNode] = None)

  private var head: Option[ListNode] = None
  private var size: Int = 0

  def insert(data: A): Unit = {
    val newNode = new ListNode(data)

    head match {
      case None =>
        head = Some(newNode)
      case Some(h) =>
        var current = h
        while (current.next.isDefined) {
          current = current.next.get
        }
        current.next = Some(newNode)
    }
    size += 1
  }

  def delete(key: A): Boolean = {
    head match {
      case None => false
      case Some(h) if h.data == key =>
        head = h.next
        size -= 1
        true
      case Some(h) =>
        var current = h
        while (current.next.isDefined && current.next.get.data != key) {
          current = current.next.get
        }
        current.next match {
          case Some(nodeToDelete) =>
            current.next = nodeToDelete.next
            size -= 1
            true
          case None => false
        }
    }
  }

  def printAll(): Unit = {
    head match {
      case None => println("List is empty")
      case Some(h) =>
        var current: Option[ListNode] = Some(h)
        while (current.isDefined) {
          print(s"${current.get.data}")
          current = current.get.next
          if (current.isDefined) print(" -> ")
        }
        println()
    }
  }

  def count: Int = size

  def printAt(pos: Int): Option[A] = {
    if (pos < 0 || pos >= size) {
      None
    } else {
      var current = head
      var index = 0
      while (index < pos && current.isDefined) {
        current = current.get.next
        index += 1
      }
      current.map(_.data)
    }
  }

  def toList: List[A] = {
    def loop(node: Option[ListNode], acc: List[A]): List[A] = node match {
      case None => acc.reverse
      case Some(n) => loop(n.next, n.data :: acc)
    }
    loop(head, Nil)
  }
}

object SinglyLinkedList {
  def apply[A](elements: A*): LinkedList[A] = {
    elements.foldRight[LinkedList[A]](Empty)((elem, list) => Node(elem, list))
  }

  def demo(): Unit = {
    println("=== Immutable LinkedList Demo ===")
    val list1 = SinglyLinkedList(1, 2, 3, 4, 5)
    print("List: ")
    list1.printAll()
    println(s"Count: ${list1.count}")
    println(s"Element at index 2: ${list1.get(2)}")

    val list2 = list1.append(6)
    print("After append(6): ")
    list2.printAll()

    val list3 = list2.delete(3)
    print("After delete(3): ")
    list3.printAll()

    println("\n=== Mutable LinkedList Demo ===")
    val mutableList = new MutableSinglyLinkedList[Long]()
    mutableList.insert(10L)
    mutableList.insert(20L)
    mutableList.insert(30L)
    mutableList.insert(40L)

    print("List: ")
    mutableList.printAll()
    println(s"Count: ${mutableList.count}")
    println(s"Element at position 2: ${mutableList.printAt(2)}")

    mutableList.delete(20L)
    print("After delete(20): ")
    mutableList.printAll()
  }
}

File: python/com/mounish/data_structures/singly_linked_list.py
"""
Singly Linked List implementations in Python with improvements.

Features:
- Type hints for better code clarity
- Multiple implementation approaches
- Pythonic idioms (iterators, generators)
- Dataclass for node representation
- Both mutable and immutable versions
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar, Optional, Iterator, List
from abc import ABC, abstractmethod

T = TypeVar('T')


@dataclass
class ListNode(Generic[T]):
    """Node in a singly linked list."""
    data: T
    next: Optional[ListNode[T]] = None

    def __repr__(self) -> str:
        return f"Node({self.data})"


class SinglyLinkedList(Generic[T]):
    """
    Mutable singly linked list implementation (similar to Java version).

    Improvements over Java:
    - Type hints for type safety
    - Pythonic methods (__len__, __iter__, __str__)
    - Generator for efficient iteration
    - Context manager support
    - Property decorators
    """

    def __init__(self) -> None:
        self._head: Optional[ListNode[T]] = None
        self._size: int = 0

    @property
    def head(self) -> Optional[ListNode[T]]:
        """Get the head node."""
        return self._head

    @property
    def size(self) -> int:
        """Get the size of the list."""
        return self._size

    def insert(self, data: T) -> None:
        """Insert a node at the end of the list."""
        new_node = ListNode(data)

        if not self._head:
            self._head = new_node
        else:
            current = self._head
            while current.next:
                current = current.next
            current.next = new_node

        self._size += 1

    def insert_at_beginning(self, data: T) -> None:
        """Insert a node at the beginning of the list."""
        new_node = ListNode(data, self._head)
        self._head = new_node
        self._size += 1

    def insert_at(self, data: T, position: int) -> bool:
        """Insert a node at a specific position."""
        if position < 0 or position > self._size:
            return False

        if position == 0:
            self.insert_at_beginning(data)
            return True

        new_node = ListNode(data)
        current = self._head
        for _ in range(position - 1):
            if current:
                current = current.next

        if current:
            new_node.next = current.next
            current.next = new_node
            self._size += 1
            return True

        return False

    def delete(self, key: T) -> bool:
        """Delete the first node with the given value."""
        if not self._head:
            return False

        if self._head.data == key:
            self._head = self._head.next
            self._size -= 1
            return True

        current = self._head
        while current.next:
            if current.next.data == key:
                current.next = current.next.next
                self._size -= 1
                return True
            current = current.next

        return False

    def delete_at(self, position: int) -> bool:
        """Delete node at specific position."""
        if position < 0 or position >= self._size or not self._head:
            return False

        if position == 0:
            self._head = self._head.next
            self._size -= 1
            return True

        current = self._head
        for _ in range(position - 1):
            if current:
                current = current.next

        if current and current.next:
            current.next = current.next.next
            self._size -= 1
            return True

        return False

    def get(self, position: int) -> Optional[T]:
        """Get value at specific position."""
        if position < 0 or position >= self._size:
            return None

        current = self._head
        for _ in range(position):
            if current:
                current = current.next

        return current.data if current else None

    def search(self, key: T) -> int:
        """Search for a value and return its position (-1 if not found)."""
        current = self._head
        position = 0

        while current:
            if current.data == key:
                return position
            current = current.next
            position += 1

        return -1

    def reverse(self) -> None:
        """Reverse the linked list in-place."""
        prev = None
        current = self._head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self._head = prev

    def to_list(self) -> List[T]:
        """Convert linked list to Python list."""
        return list(self)

    def __len__(self) -> int:
        """Return the length of the list."""
        return self._size

    def __iter__(self) -> Iterator[T]:
        """Iterate over the list values."""
        current = self._head
        while current:
            yield current.data
            current = current.next

    def __str__(self) -> str:
        """String representation of the list."""
        if not self._head:
            return "[]"
        return " -> ".join(str(data) for data in self)

    def __repr__(self) -> str:
        """Detailed representation."""
        return f"SinglyLinkedList([{', '.join(repr(data) for data in self)}])"

    def __contains__(self, item: T) -> bool:
        """Check if item exists in the list."""
        return self.search(item) != -1

    def __getitem__(self, index: int) -> T:
        """Get item by index (supports negative indexing)."""
        if index < 0:
            index = self._size + index

        value = self.get(index)
        if value is None:
            raise IndexError("list index out of range")
        return value


class ImmutableLinkedList(Generic[T], ABC):
    """Abstract base class for immutable linked list."""

    @abstractmethod
    def is_empty(self) -> bool:
        """Check if list is empty."""
        pass

    @abstractmethod
    def head(self) -> T:
        """Get the head value."""
        pass

    @abstractmethod
    def tail(self) -> ImmutableLinkedList[T]:
        """Get the tail of the list."""
        pass

    @abstractmethod
    def prepend(self, value: T) -> ImmutableLinkedList[T]:
        """Prepend a value to the list."""
        pass

    @abstractmethod
    def append(self, value: T) -> ImmutableLinkedList[T]:
        """Append a value to the list."""
        pass


class Empty(ImmutableLinkedList[T]):
    """Empty immutable linked list."""

    def is_empty(self) -> bool:
        return True

    def head(self) -> T:
        raise IndexError("head of empty list")

    def tail(self) -> ImmutableLinkedList[T]:
        raise IndexError("tail of empty list")

    def prepend(self, value: T) -> ImmutableLinkedList[T]:
        return Cons(value, self)

    def append(self, value: T) -> ImmutableLinkedList[T]:
        return Cons(value, self)

    def __str__(self) -> str:
        return "[]"

    def __repr__(self) -> str:
        return "Empty()"


class Cons(ImmutableLinkedList[T]):
    """Non-empty immutable linked list (cons cell)."""

    def __init__(self, head_val: T, tail_list: ImmutableLinkedList[T]):
        self._head = head_val
        self._tail = tail_list

    def is_empty(self) -> bool:
        return False

    def head(self) -> T:
        return self._head

    def tail(self) -> ImmutableLinkedList[T]:
        return self._tail

    def prepend(self, value: T) -> ImmutableLinkedList[T]:
        return Cons(value, self)

    def append(self, value: T) -> ImmutableLinkedList[T]:
        if self._tail.is_empty():
            return Cons(self._head, Cons(value, Empty()))
        return Cons(self._head, self._tail.append(value))

    def __str__(self) -> str:
        result = [str(self._head)]
        current = self._tail
        while not current.is_empty():
            result.append(str(current.head()))
            current = current.tail()
        return " -> ".join(result)

    def __repr__(self) -> str:
        return f"Cons({self._head}, {self._tail})"


def create_linked_list(*values: T) -> ImmutableLinkedList[T]:
    """Create an immutable linked list from values."""
    result: ImmutableLinkedList[T] = Empty()
    for value in reversed(values):
        result = result.prepend(value)
    return result


def demo() -> None:
    """Demonstrate linked list implementations."""
    print("=" * 80)
    print("SINGLY LINKED LIST DEMO")
    print("=" * 80)

    # Mutable linked list demo
    print("\n=== Mutable Linked List ===")
    mutable_list = SinglyLinkedList[int]()

    print("\nInserting values: 10, 20, 30, 40, 50")
    for value in [10, 20, 30, 40, 50]:
        mutable_list.insert(value)

    print(f"List: {mutable_list}")
    print(f"Length: {len(mutable_list)}")
    print(f"As Python list: {mutable_list.to_list()}")

    print("\nInserting 5 at beginning:")
    mutable_list.insert_at_beginning(5)
    print(f"List: {mutable_list}")

    print("\nInserting 25 at position 4:")
    mutable_list.insert_at(25, 4)
    print(f"List: {mutable_list}")

    print(f"\nGet element at position 3: {mutable_list.get(3)}")
    print(f"Search for 30: position {mutable_list.search(30)}")
    print(f"Contains 30: {30 in mutable_list}")
    print(f"Contains 100: {100 in mutable_list}")

    print("\nDeleting 20:")
    mutable_list.delete(20)
    print(f"List: {mutable_list}")

    print("\nDeleting at position 2:")
    mutable_list.delete_at(2)
    print(f"List: {mutable_list}")

    print("\nReversing the list:")
    mutable_list.reverse()
    print(f"List: {mutable_list}")

    print("\nIterating with for loop:")
    for i, value in enumerate(mutable_list):
        print(f"  Position {i}: {value}")

    # Immutable linked list demo
    print("\n=== Immutable Linked List ===")
    immutable_list = create_linked_list(1, 2, 3, 4, 5)
    print(f"Created list: {immutable_list}")

    print(f"\nHead: {immutable_list.head()}")
    print(f"Tail: {immutable_list.tail()}")

    list2 = immutable_list.prepend(0)
    print(f"\nAfter prepend(0): {list2}")
    print(f"Original list: {immutable_list}")

    list3 = immutable_list.append(6)
    print(f"\nAfter append(6): {list3}")
    print(f"Original list: {immutable_list}")

    # Using indexing
    print("\n=== Indexing Support ===")
    test_list = SinglyLinkedList[str]()
    for word in ["Python", "is", "awesome"]:
        test_list.insert(word)

    print(f"List: {test_list}")
    print(f"list[0] = {test_list[0]}")
    print(f"list[1] = {test_list[1]}")
    print(f"list[-1] = {test_list[-1]}")

    print("\n" + "=" * 80)


if __name__ == "__main__":
    demo()
